### PR TITLE
Implemented progress bars for classifier training

### PR DESCRIPTION
**Pull request by vijay120**
This PR implements progress bars for MindMeld classifiers, solving issue https://github.com/cisco/mindmeld/issues/200. The current implementation makes a few assumptions:
1. Each classifier takes the same time to train
2. Doesn't take into account data loading, which can take awhile.

### CODE REVIEW

File: mindmeld/components/nlp.py [hunk 1]
```diff
@@ -421,10 +422,11 @@ def __init__(self, app_path, resource_loader=None, config=None):
         self.name = app_path
         self._load_custom_features()
         self.domain_classifier = DomainClassifier(self.resource_loader)
+        self.progress_bar = progress_bar or tqdm(total=0, desc="Classifier progress")
```
Comment (minhtuev) [inline mindmeld/components/nlp.py:425]: Do we have a single progress bar for the entire training process, or an individual progress bar for each domain?

File: mindmeld/components/nlp.py [hunk 2]
```diff
@@ -770,9 +777,17 @@ def __init__(self, app_path, domain, resource_loader=None):
...
         self.name = domain
         self.intent_classifier = IntentClassifier(self.resource_loader, domain)
-        for intent in path.get_intents(app_path, domain):
+        intents = path.get_intents(app_path, domain)
+
+        # If there is only one intent in the domain, the classifier would not run
+        # hence we only account for classifiers were there are two or more intents
+        if len(intents) > 1:
+            self.progress_bar = progress_bar
+            self.progress_bar.total += 1
```
Comment (minhtuev) [inline mindmeld/components/nlp.py:786]: We should increase total by the number of intents right?

File: mindmeld/components/nlp.py (hunk 1 above)
Comment (vijay120) [reply mindmeld/components/nlp.py:425]: One progress bar for all the classifiers

File: mindmeld/components/nlp.py (hunk 2 above)
Comment (vijay120) [reply mindmeld/components/nlp.py:786]: We dont technically train a classifier if there is only 1 intent.

File: mindmeld/components/nlp.py (hunk 1 above)
Comment (minhtuev) [reply mindmeld/components/nlp.py:425]: Should we have an option to disable the progress bar?

**Comment by minhtuev** [comment]
This is my standard output using the progress bar:
```
/Users/minhtue/opt/anaconda3/lib/python3.7/site-packages/sklearn/feature_extraction/text.py:17: DeprecationWarning: Using or importing the ABCs from 'collections' instead of from 'collections.abc' is deprecated since Python 3.3,and in 3.9 it will stop working
  from collections import Mapping, defaultdict
/Users/minhtue/opt/anaconda3/lib/python3.7/site-packages/sklearn/ensemble/weight_boosting.py:29: DeprecationWarning: numpy.core.umath_tests is an internal NumPy module and should not be imported. It will be removed in a future NumPy release.
  from numpy.core.umath_tests import inner1d
Numerical parser running, PID 76921
warning: The application package kwik_e_mart is already imported.
Classifier progress: 0it [00:00, ?it/s]No question_answering model configuration set. Using default.
Classifier progress:   0%|                                                                                                                                                                                                                                                                                                               | 0/7 [00:00<?, ?it/s]Fitting intent classifier: domain='store_info'
Loading raw queries from file /Users/minhtue/workspace/mindmeld/tests/kwik_e_mart/domains/store_info/exit/train.txt
Loading raw queries from file /Users/minhtue/workspace/mindmeld/tests/kwik_e_mart/domains/store_info/find_nearest_store/train.txt
Loading raw queries from file /Users/minhtue/workspace/mindmeld/tests/kwik_e_mart/domains/store_info/get_store_hours/train.txt
Loading raw queries from file /Users/minhtue/workspace/mindmeld/tests/kwik_e_mart/domains/store_info/get_store_number/train.txt
Loading raw queries from file /Users/minhtue/workspace/mindmeld/tests/kwik_e_mart/domains/store_info/greet/train.txt
Loading raw queries from file /Users/minhtue/workspace/mindmeld/tests/kwik_e_mart/domains/store_info/help/train.txt
Loading queries from file /Users/minhtue/workspace/mindmeld/tests/kwik_e_mart/domains/store_info/exit/train.txt
Loading queries from file /Users/minhtue/workspace/mindmeld/tests/kwik_e_mart/domains/store_info/find_nearest_store/train.txt
Loading queries from file /Users/minhtue/workspace/mindmeld/tests/kwik_e_mart/domains/store_info/get_store_hours/train.txt
Loading queries from file /Users/minhtue/workspace/mindmeld/tests/kwik_e_mart/domains/store_info/get_store_number/train.txt
Loading queries from file /Users/minhtue/workspace/mindmeld/tests/kwik_e_mart/domains/store_info/greet/train.txt
Loading queries from file /Users/minhtue/workspace/mindmeld/tests/kwik_e_mart/domains/store_info/help/train.txt
Selecting hyperparameters using k-fold cross-validation with 10 splits
Best accuracy: 97.34%, params: {'C': 100, 'class_weight': {0: 1.1006451612903225, 1: 1.2495575221238937, 2: 0.9003225806451612, 3: 1.7525423728813558, 4: 0.8220039292730844, 5: 1.3468749999999998}, 'fit_intercept': False}
Classifier progress:  14%|██████████████████████████████████████████▏                                                                                                                                                                                                                                                            | 1/7 [00:08<00:49,  8.25s/it]Fitting entity recognizer: domain='store_info', intent='help'
There are no labels in this label set, so we don't fit the model.
Classifier progress:  29%|████████████████████████████████████████████████████████████████████████████████████▎                                                                                                                                                                                                                  | 2/7 [00:08<00:41,  8.25s/it]Fitting entity recognizer: domain='store_info', intent='greet'
There are no labels in this label set, so we don't fit the model.
Classifier progress:  43%|██████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████▍                                                                                                                                                                        | 3/7 [00:08<00:33,  8.25s/it]Fitting entity recognizer: domain='store_info', intent='get_store_number'
Selecting hyperparameters using k-fold cross-validation with 5 splits
Best accuracy: 98.60%, params: {'C': 1000000, 'penalty': 'l1'}
Classifier progress:  57%|████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████▌                                                                                                                              | 4/7 [00:08<00:17,  5.83s/it]No entity_resolution model configuration set. Using default.
Fitting role classifier: domain='store_info', intent='get_store_number', entity_type='store_name'
No role model configuration set. Using default.
Importing synonym data to synonym index 'synonym_store_name'
warning: Elasticsearch index 'synonym_store_name' for application 'kwik_e_mart' already exists!
Loading index 'synonym_store_name'
100%|██████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 25/25 [00:01<00:00, 13.97it/s]
Loaded 25 documents                                                                                                                                                                                                                                                                                                             | 1/25 [00:01<00:42,  1.79s/it]
Classifier progress:  62%|████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████▍                                                                                                              | 5/8 [00:12<00:15,  5.19s/it]Fitting entity recognizer: domain='store_info', intent='get_store_hours'
Loading raw queries from file /Users/minhtue/workspace/mindmeld/tests/kwik_e_mart/domains/store_info/get_store_hours/testtrain123.txt
Loading queries from file /Users/minhtue/workspace/mindmeld/tests/kwik_e_mart/domains/store_info/get_store_hours/testtrain123.txt
Classifier progress:  75%|█████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████▎                                                                         | 6/8 [00:12<00:07,  3.76s/it]No entity_resolution model configuration set. Using default.
No entity_resolution model configuration set. Using default.
Fitting role classifier: domain='store_info', intent='get_store_hours', entity_type='store_name'
No role model configuration set. Using default.
Importing synonym data to synonym index 'synonym_store_name'
warning: Elasticsearch index 'synonym_store_name' for application 'kwik_e_mart' already exists!
Loading index 'synonym_store_name'
100%|██████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 25/25 [00:00<00:00, 27.90it/s]
Loaded 25 documents                                                                                                                                                                                                                                                                                                             | 1/25 [00:00<00:21,  1.12it/s]
Classifier progress:  70%|█████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████▊                                                                                        | 7/10 [00:14<00:09,  3.06s/it]Fitting role classifier: domain='store_info', intent='get_store_hours', entity_type='sys_time'
No role model configuration set. Using default.
Classifier progress:  80%|███████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████▏                                                          | 8/10 [00:14<00:06,  3.06s/it]Fitting entity recognizer: domain='store_info', intent='exit'
There are no labels in this label set, so we don't fit the model.
Classifier progress:  90%|████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████▌                             | 9/10 [00:14<00:03,  3.06s/it]Fitting entity recognizer: domain='store_info', intent='find_nearest_store'
There are no labels in this label set, so we don't fit the model.
Classifier progress: 100%|█████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 10/10 [00:14<00:00,  3.06s/it]Saving intent classifier: domain='store_info'
Saving entity classifier: domain='store_info', intent='help'
Saving entity classifier: domain='store_info', intent='greet'
Saving entity classifier: domain='store_info', intent='get_store_number'
Saving role classifier: domain='store_info', intent='get_store_number', entity='store_name'
Saving entity classifier: domain='store_info', intent='get_store_hours'
Saving role classifier: domain='store_info', intent='get_store_hours', entity='store_name'
Saving role classifier: domain='store_info', intent='get_store_hours', entity='sys_time'
Saving entity classifier: domain='store_info', intent='exit'
Saving entity classifier: domain='store_info', intent='find_nearest_store'
Classifier progress: 100%|█████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 10/10 [00:14<00:00,  1.46s/it]

**Comment by minhtuev** [comment]
I think the output is a bit confusing, so we might consider these options:
1) We should name each progress bar individually instead of `Classifier progress`
2) We should disable other standard outputs while we are using the progress bars
3) Alternatively, we can just have one progress bar for the entire process (and disabling the rest of the standard output).

**Comment by vijay120** [comment]
The reason why you are getting that output is because all of MM logging to going to standard out, which is inevitable if you configured the logger as such. You can prevent that from happening by doing this:
```
In [4]: from mindmeld.components.nlp import NaturalLanguageProcessor   
   ...: import mindmeld as mm 
   ...: import logging   
   ...: mm.configure_logs(level=logging.ERROR)  
   ...: nlp = NaturalLanguageProcessor(app_path='banking_assistant') 
   ...: nlp.build() 
```

If you do the above, this is the output you will get:
```
Classifier progress: 0it [00:00, ?it/s]
Classifier progress:   0%|          | 0/20 [00:00<?, ?it/s]
Classifier progress:   5%|▌         | 1/20 [00:28<09:01, 28.49s/it]
Classifier progress:   5%|▌         | 1/20 [00:28<09:01, 28.49s/it]
Classifier progress:  10%|█         | 2/20 [00:29<06:04, 20.25s/it]
Classifier progress:  10%|█         | 2/20 [00:29<06:04, 20.25s/it]
Classifier progress:  15%|█▌        | 3/20 [00:29<05:44, 20.25s/it]
Classifier progress:  20%|██        | 4/20 [00:29<05:23, 20.25s/it]
Classifier progress:  25%|██▌       | 5/20 [00:31<03:35, 14.38s/it]
Classifier progress:  25%|██▌       | 5/20 [00:31<03:35, 14.38s/it]
Classifier progress:  30%|███       | 6/20 [00:31<03:21, 14.38s/it]
Classifier progress:  35%|███▌      | 7/20 [00:31<03:06, 14.38s/it]
Classifier progress:  40%|████      | 8/20 [00:31<02:52, 14.38s/it]
Classifier progress:  45%|████▌     | 9/20 [00:31<02:38, 14.38s/it]
Classifier progress:  50%|█████     | 10/20 [00:31<02:23, 14.38s/it]
Classifier progress:  55%|█████▌    | 11/20 [00:31<02:09, 14.38s/it]
Classifier progress:  60%|██████    | 12/20 [00:31<01:55, 14.38s/it]
Classifier progress:  65%|██████▌   | 13/20 [00:31<01:40, 14.38s/it]
Classifier progress:  70%|███████   | 14/20 [00:31<01:26, 14.38s/it]
Classifier progress:  75%|███████▌  | 15/20 [00:33<00:50, 10.11s/it]
Classifier progress:  75%|███████▌  | 15/20 [00:33<00:50, 10.11s/it]
Classifier progress:  80%|████████  | 16/20 [00:33<00:29,  7.28s/it]
Classifier progress:  80%|████████  | 16/20 [00:33<00:29,  7.28s/it]/Users/vijay/.pyenv/versions/3.6.5/envs/mindmeld/lib/python3.6/site-packages/elasticsearch/connection/base.py:177: ElasticsearchDeprecationWarning: [types removal] The parameter include_type_name should be explicitly specified in get indices requests to prepare for 7.0. In 7.0 include_type_name will default to 'false', which means responses will omit the type name in mapping definitions.
  warnings.warn(message, category=ElasticsearchDeprecationWarning)


  0%|          | 0/1 [00:00<?, ?it/s]

100%|██████████| 1/1 [00:00<00:00,  6.98it/s]

Classifier progress:  81%|████████  | 17/21 [00:34<00:21,  5.28s/it]
Classifier progress:  81%|████████  | 17/21 [00:34<00:21,  5.28s/it]
Classifier progress:  86%|████████▌ | 18/21 [00:35<00:11,  3.90s/it]
Classifier progress:  86%|████████▌ | 18/21 [00:35<00:11,  3.90s/it]/Users/vijay/.pyenv/versions/3.6.5/envs/mindmeld/lib/python3.6/site-packages/elasticsearch/connection/base.py:177: ElasticsearchDeprecationWarning: [types removal] The parameter include_type_name should be explicitly specified in get indices requests to prepare for 7.0. In 7.0 include_type_name will default to 'false', which means responses will omit the type name in mapping definitions.
  warnings.warn(message, category=ElasticsearchDeprecationWarning)


100%|██████████| 3/3 [00:00<00:00, 61.19it/s]

Classifier progress:  83%|████████▎ | 19/23 [00:35<00:11,  2.77s/it]
Classifier progress:  83%|████████▎ | 19/23 [00:35<00:11,  2.77s/it]
Classifier progress:  87%|████████▋ | 20/23 [00:35<00:08,  2.77s/it]
Classifier progress:  91%|█████████▏| 21/23 [00:35<00:04,  2.04s/it]
Classifier progress:  91%|█████████▏| 21/23 [00:35<00:04,  2.04s/it]/Users/vijay/.pyenv/versions/3.6.5/envs/mindmeld/lib/python3.6/site-packages/elasticsearch/connection/base.py:177: ElasticsearchDeprecationWarning: [types removal] The parameter include_type_name should be explicitly specified in get indices requests to prepare for 7.0. In 7.0 include_type_name will default to 'false', which means responses will omit the type name in mapping definitions.
  warnings.warn(message, category=ElasticsearchDeprecationWarning)


100%|██████████| 3/3 [00:00<00:00, 46.16it/s]

Classifier progress:  92%|█████████▏| 22/24 [00:36<00:02,  1.47s/it]
Classifier progress:  92%|█████████▏| 22/24 [00:36<00:02,  1.47s/it]
Classifier progress:  96%|█████████▌| 23/24 [00:37<00:01,  1.32s/it]
Classifier progress:  96%|█████████▌| 23/24 [00:37<00:01,  1.32s/it]/Users/vijay/.pyenv/versions/3.6.5/envs/mindmeld/lib/python3.6/site-packages/elasticsearch/connection/base.py:177: ElasticsearchDeprecationWarning: [types removal] The parameter include_type_name should be explicitly specified in get indices requests to prepare for 7.0. In 7.0 include_type_name will default to 'false', which means responses will omit the type name in mapping definitions.
  warnings.warn(message, category=ElasticsearchDeprecationWarning)


100%|██████████| 2/2 [00:00<00:00, 33.86it/s]

Classifier progress:  92%|█████████▏| 24/26 [00:37<00:01,  1.04it/s]
Classifier progress:  92%|█████████▏| 24/26 [00:37<00:01,  1.04it/s]
Classifier progress:  96%|█████████▌| 25/26 [00:37<00:00,  1.04it/s]
Classifier progress: 100%|██████████| 26/26 [00:37<00:00,  1.04it/s]
```

The client can also listen to these updates by passing their own tqdm progress bar to the NLP object.

>> 1.We should name each progress bar individually instead of Classifier progress

Like I said, there is only one progress bar here. The intention of this feature is to give a consolidate progress of the ML training in a MindMeld application. Having a progress bar for each classifier (the banking assistant has 26), defeats the purpose of having a consolidate UI for classifier training progress.


>> 2.We should disable other standard outputs while we are using the progress bars

The client can do it using the `configure_logs` feature.

>> Alternatively, we can just have one progress bar for the entire process (and disabling the rest of the standard output).

This is what is implemented.

**Comment by minhtuev** [comment]
@vijay120: I was just running the command `python -m [blueprint] build`, which I think a lot of people will do by default, so we should configure the logs not to output when we are displaying the progress bar.

If you only want one progress bar, you should update the field "total" from the beginning with all the intents, otherwise you will have issues like this (see above):
```
Classifier progress:  81%|████████  | 17/21 [00:34<00:21,  5.28s/it]
Classifier progress:  81%|████████  | 17/21 [00:34<00:21,  5.28s/it]
Classifier progress:  86%|████████▌ | 18/21 [00:35<00:11,  3.90s/it]
Classifier progress:  86%|████████▌ | 18/21 [00:35<00:11,  3.90s/it]
Classifier progress:  83%|████████▎ | 19/23 [00:35<00:11,  2.77s/it]
Classifier progress:  83%|████████▎ | 19/23 [00:35<00:11,  2.77s/it]
Classifier progress:  87%|████████▋ | 20/23 [00:35<00:08,  2.77s/it]
Classifier progress:  91%|█████████▏| 21/23 [00:35<00:04,  2.04s/it]
Classifier progress:  91%|█████████▏| 21/23 [00:35<00:04,  2.04s/it]
```
Here the progress seems to have regressed from 86% to 83% even though we are just updating the total value.

**Comment by vijay120** [comment]
@minhtuev I don't have any solutions for getting it to work on `python -m [blueprint] build` without removing all our existing logging statements. It will be noisy unless the existing statements are removed. The main usecase here was for Botlite to subscribe to events from the progress bar, so it's not intended for `python -m [blueprint] build` type usecases. Do you have any suggestions on how we can do it then?

Regarding the increase of the total, the reason why this is happening is because `IntentProcessor` class ONLY initializes its children in the build command:
```
    def _build(self, incremental=False, label_set=None):
        """Builds the models for this intent"""

        # train entity recognizer
        self.entity_recognizer.fit(
            label_set=label_set, incremental_timestamp=self.incremental_timestamp
        )

        self.progress_bar.update(1)
        self.progress_bar.refresh()

        # Create the entity processors
        entity_types = self.entity_recognizer.entity_types
        for entity_type in entity_types:
            processor = EntityProcessor(
                self._app_path,
                self.domain,
                self.name,
                entity_type,
                self.resource_loader,
                self.progress_bar
            )
            self._children[entity_type] = processor
```

So there is no way to know how many children the IntentProcessor have until calling _build, which happens only after the parent classifiers are built. Hence, the change in the total. I think this is fine since I see such behavior in other progress bars as well and the Entity recognizers are generally small. Happy to take recommendations on how to fix this as well.

File: mindmeld/components/nlp.py [hunk 3]
```diff
@@ -459,7 +460,8 @@ def _build(self, incremental=False, label_set=None):
 
         # reset display for the progress bar. This is important for repeated use of the
         # progress bar
-        self.progress_bar.reset()
+        if self.progress_bar is not None:
```
Comment (minhtuev) [inline mindmeld/components/nlp.py:463]:
```suggestion
        if self.progress_bar:
```

File: mindmeld/components/nlp.py [hunk 4]
```diff
@@ -781,8 +785,8 @@ def __init__(self, app_path, domain, resource_loader=None, progress_bar=None):
 
         # If there is only one intent in the domain, the classifier would not run
         # hence we only account for classifiers were there are two or more intents
-        if len(intents) > 1:
-            self.progress_bar = progress_bar
+        self.progress_bar = progress_bar
+        if len(intents) > 1 and self.progress_bar is not None:
```
Comment (minhtuev) [inline mindmeld/components/nlp.py:789]:
```suggestion
        if len(intents) > 1 and self.progress_bar:
```

File: mindmeld/components/nlp.py [hunk 5]
```diff
@@ -798,7 +802,7 @@ def _build(self, incremental=False, label_set=None):
             label_set=label_set, incremental_timestamp=self.incremental_timestamp
         )
 
-        if len(self._children) > 1:
+        if len(self._children) > 1 and self.progress_bar is not None:
```
Comment (minhtuev) [inline mindmeld/components/nlp.py:805]: same

File: mindmeld/components/nlp.py (hunk 4 above)
Comment (vijay120) [reply mindmeld/components/nlp.py:789]: I cannot do this is `tqdm` objects pass that check but are not None objects ie when self.progress_bar exists,  `if self.progress_bar` returns False but `if self.progress_bar is not None` return True.

File: source/userguide/nlp.rst [hunk 6]
```diff
@@ -789,3 +789,27 @@ The saved models can then be loaded anytime using the :meth:`NaturalLanguageProc
    ...
 
 Another option is to save just one specific NLP model, which is useful when you are actively experimenting with individual classifiers and want to checkpoint your work or save multiple model versions for comparison. This is done using the :meth:`dump` and :meth:`load` methods exposed by each classifier. Refer to the chapter for the appropriate classifier to learn more.
+
+
+Tracking classifier progress
+----------------------------
+
+Training classifiers can be time-consuming, so progress tracking is important for client's to see how much longer the application will finis
```
Comment (vembar) [inline source/userguide/nlp.rst:797]: Consider rewording to something like this?

Training the NLU pipeline can be time-consuming depending on the number of models to be trained and the number of training examples for each of them. The following code snippet shows how you can do progress tracking   while building your application. The progress bar will display the fraction of models that have been trained at any given time.

**Comment by arushir** [comment]
I agree with Varsha that we should update the name to `Models trained` or `NLP progress`. 

I'm wondering if there is a way to flush the output so that the progress bar updates in place? That would definitely look cleaner. This might be worth looking in to: https://stackoverflow.com/questions/45862715/how-to-flush-tqdm-progress-bar-explicitly

File: source/userguide/nlp.rst (hunk 6 above)
Comment (vijay120) [reply source/userguide/nlp.rst:797]: done

**Comment by vijay120** [comment]
>> I'm wondering if there is a way to flush the output so that the progress bar updates in place?

That is an artifact of Jupyter notebooks + other logging statements that mess up the progress. Here is how it looks in `ipython`:

```
In [3]:    import mindmeld 
   ...:    import logging 
   ...:    from mindmeld.components.nlp import NaturalLanguageProcessor 
   ...:    from tqdm import tqdm 
   ...:  
   ...:    # Setting the logging level to ERROR prevents the tqdm progress bar from interleaving 
   ...:    # with other logging messages that are written to stdout. 
   ...:    mindmeld.configure_logs(level=logging.ERROR) 
   ...:  
   ...:    # Make sure the total parameter starts at 0 since the classifiers will increment the total as 
   ...:    # they get initialized 
   ...:    pbar = tqdm(total=0, desc="NLP progress") 
   ...:    nlp = NaturalLanguageProcessor(app_path='banking_assistant', progress_bar=pbar) 
   ...:    nlp.build()                                                                                                                                        

/Users/vijay/.pyenv/versions/3.6.5/envs/mindmeld/lib/python3.6/site-packages/sklearn/ensemble/weight_boosting.py:29: DeprecationWarning: numpy.core.umath_tests is an internal NumPy module and should not be imported. It will be removed in a future NumPy release.
  from numpy.core.umath_tests import inner1d
NLP progress:  77%|██████████████████████████████████████████████████████████████████████████████████▋                        | 17/22 [00:31<00:49,  9.90s/it]/Users/vijay/.pyenv/versions/3.6.5/envs/mindmeld/lib/python3.6/site-packages/elasticsearch/connection/base.py:177: ElasticsearchDeprecationWarning: [types removal] The parameter include_type_name should be explicitly specified in get indices requests to prepare for 7.0. In 7.0 include_type_name will default to 'false', which means responses will omit the type name in mapping definitions.
  warnings.warn(message, category=ElasticsearchDeprecationWarning)
100%|███████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 2/2 [00:00<00:00, 26.74it/s]
NLP progress:  91%|█████████████████████████████████████████████████████████████████████████████████████████████████▎         | 20/22 [00:32<00:09,  4.96s/it]/Users/vijay/.pyenv/versions/3.6.5/envs/mindmeld/lib/python3.6/site-packages/elasticsearch/connection/base.py:177: ElasticsearchDeprecationWarning: [types removal] The parameter include_type_name should be explicitly specified in get indices requests to prepare for 7.0. In 7.0 include_type_name will default to 'false', which means responses will omit the type name in mapping definitions.
  warnings.warn(message, category=ElasticsearchDeprecationWarning)
100%|███████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 1/1 [00:00<00:00, 29.54it/s]
NLP progress:  96%|██████████████████████████████████████████████████████████████████████████████████████████████████████▎    | 22/23 [00:33<00:03,  3.58s/it]/Users/vijay/.pyenv/versions/3.6.5/envs/mindmeld/lib/python3.6/site-packages/elasticsearch/connection/base.py:177: ElasticsearchDeprecationWarning: [types removal] The parameter include_type_name should be explicitly specified in get indices requests to prepare for 7.0. In 7.0 include_type_name will default to 'false', which means responses will omit the type name in mapping definitions.
  warnings.warn(message, category=ElasticsearchDeprecationWarning)
100%|███████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 3/3 [00:00<00:00, 59.19it/s]
NLP progress: 100%|███████████████████████████████████████████████████████████████████████████████████████████████████████████| 24/24 [00:33<00:00,  2.52s/it]
```

If we didnt have the `ElasticsearchDeprecationWarning` warnings, the progress bar would be a one liner. But since those deprecation warnings happen, it has to be shown, splitting up the pbar.

**Comment by arushir** [comment]
@vijay120 It has the same printing behavior outside of Jupyter notebook as well, right? If someone runs `build` from the command line it would be nice to have a cleaner output.

I understand that the progress bar is interrupted by the print statements. The [link I posted](https://stackoverflow.com/questions/45862715/how-to-flush-tqdm-progress-bar-explicitly) suggests that we can get around that in some ways. Did you look into those options?

**Comment by vijay120** [comment]
So the ES warnings are written to stderr and tqdm is also written by default to stderr, so they should work according to the link you posted. But I anyways forced the write to stderr, but the results were the same:

```
In [1]:    import mindmeld 
   ...:    import logging 
   ...:    from mindmeld.components.nlp import NaturalLanguageProcessor 
   ...:    from tqdm import tqdm 
   ...:    import sys 
   ...:    # Setting the logging level to ERROR prevents the tqdm progress bar from interleaving 
   ...:    # with other logging messages that are written to stdout. 
   ...:    mindmeld.configure_logs(level=logging.ERROR) 
   ...:  
   ...:    # Make sure the total parameter starts at 0 since the classifiers will increment the total as 
   ...:    # they get initialized 
   ...:    pbar = tqdm(total=0, file=sys.stderr, desc="NLP progress") 
   ...:    nlp = NaturalLanguageProcessor(app_path='banking_assistant', progress_bar=pbar) 
   ...:    nlp.build()                                                                                                                                        
/Users/vijay/.pyenv/versions/3.6.5/envs/mindmeld/lib/python3.6/site-packages/sklearn/ensemble/weight_boosting.py:29: DeprecationWarning: numpy.core.umath_tests is an internal NumPy module and should not be imported. It will be removed in a future NumPy release.
  from numpy.core.umath_tests import inner1d
NLP progress:  10%|██████████▊                                                                                                 | 2/20 [00:28<05:55, 19.76s/it]/Users/vijay/.pyenv/versions/3.6.5/envs/mindmeld/lib/python3.6/site-packages/elasticsearch/connection/base.py:177: ElasticsearchDeprecationWarning: [types removal] The parameter include_type_name should be explicitly specified in get indices requests to prepare for 7.0. In 7.0 include_type_name will default to 'false', which means responses will omit the type name in mapping definitions.
  warnings.warn(message, category=ElasticsearchDeprecationWarning)
100%|███████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 1/1 [00:00<00:00, 24.80it/s]
NLP progress:  19%|████████████████████▌                                                                                       | 4/21 [00:29<03:57, 13.95s/it]/Users/vijay/.pyenv/versions/3.6.5/envs/mindmeld/lib/python3.6/site-packages/elasticsearch/connection/base.py:177: ElasticsearchDeprecationWarning: [types removal] The parameter include_type_name should be explicitly specified in get indices requests to prepare for 7.0. In 7.0 include_type_name will default to 'false', which means responses will omit the type name in mapping definitions.
  warnings.warn(message, category=ElasticsearchDeprecationWarning)
100%|███████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 2/2 [00:00<00:00, 39.75it/s]
NLP progress:  39%|██████████████████████████████████████████▎                                                                 | 9/23 [00:30<01:36,  6.90s/it]/Users/vijay/.pyenv/versions/3.6.5/envs/mindmeld/lib/python3.6/site-packages/elasticsearch/connection/base.py:177: ElasticsearchDeprecationWarning: [types removal] The parameter include_type_name should be explicitly specified in get indices requests to prepare for 7.0. In 7.0 include_type_name will default to 'false', which means responses will omit the type name in mapping definitions.
  warnings.warn(message, category=ElasticsearchDeprecationWarning)
100%|███████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 3/3 [00:00<00:00, 95.20it/s]
NLP progress: 100%|███████████████████████████████████████████████████████████████████████████████████████████████████████████| 24/24 [00:32<00:00,  3.60s/it]
```

**Comment by vijay120** [comment]
If I comment out the deprecation warning in ES python code, the results are much nicer:
```
In [1]:    import mindmeld 
   ...:    import logging 
   ...:    from mindmeld.components.nlp import NaturalLanguageProcessor 
   ...:    from tqdm import tqdm 
   ...:    import sys 
   ...:    # Setting the logging level to ERROR prevents the tqdm progress bar from interleaving 
   ...:    # with other logging messages that are written to stdout. 
   ...:    mindmeld.configure_logs(level=logging.ERROR) 
   ...:  
   ...:    # Make sure the total parameter starts at 0 since the classifiers will increment the total as 
   ...:    # they get initialized 
   ...:    pbar = tqdm(total=0, file=sys.stderr, desc="NLP progress") 
   ...:    nlp = NaturalLanguageProcessor(app_path='banking_assistant', progress_bar=pbar) 
   ...:    nlp.build()                                                                                                                                        
/Users/vijay/.pyenv/versions/3.6.5/envs/mindmeld/lib/python3.6/site-packages/sklearn/ensemble/weight_boosting.py:29: DeprecationWarning: numpy.core.umath_tests is an internal NumPy module and should not be imported. It will be removed in a future NumPy release.
  from numpy.core.umath_tests import inner1d
100%|███████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 2/2 [00:00<00:00, 21.46it/s]
100%|███████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 1/1 [00:00<00:00,  6.75it/s]
100%|███████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 3/3 [00:00<00:00, 28.30it/s]
NLP progress: 100%|███████████████████████████████████████████████████████████████████████████████████████████████████████████| 24/24 [00:30<00:00,  1.42s/it]
```

**Comment by vijay120** [comment]
@arushir  Can we fix this warning: `ElasticsearchDeprecationWarning: [types removal] The parameter include_type_name should be explicitly specified in get indices requests to prepare for 7.0. In 7.0 include_type_name will default to 'false', which means responses will omit the type name in mapping definitions.`?

**Review by arushir — APPROVED** [review]

**Pull request merged**